Add locked to rollingstock model to prevent its edition/deletion

Erashin opened the pull request:
editoast:
   - RollingStock => RollingStockCommon containing only the necessary fields for Core. RollingStock now contains a RollingStockCommon and the necessary fields for front.
   - add locked to rolling_stock && light_rolling_stock models.
   - add query param to POST to set initial rolling_stock locked value. Add control to PATCH && DELETE rolling_stock endpoints to prevent it when locked is true. Add endpoint to edit the locked value of a rolling_stock.

python: api/osrd_schemas: add locked to rolling_stock model.

Closes #4009.

flomonster commented on editoast/src/tests/example_rolling_stock_4.json:
This is not needed, you can use an existing rolling stock

flomonster commented on editoast/src/schema/rolling_stock/mod.rs:
```diff
@@ -12,6 +12,7 @@ pub struct RollingStock {
     pub id: i64,
     pub name: String,
     pub version: String,
+    pub locked: bool,
```
This field shouldn't be part of the schema. It's only part of the model. Meaning we don't need to add the lock `field` to our examples.

Erashin replied:
I'm down to do this, but that means that in the responses returned by the post and get queries, we do not return the value of the rolling_stock's locked field.
It also means that in the RollingStock => RollingStockModel conversion (see views/rolling_stock.rs), we do not know what the locked field's value is. I'll put `locked = None` in this conversion (rather than `false`), so an error is thrown in case you don't specify the locked field post-conversion and try to save it to the DB (that way, if the locked field was true, it won't be overwritten without us noticing). Is that fine with you ?

Erashin replied:
New strategy: 2 schemas. RollingStock for core, without locked and metadata. RollingStockFull for front with locked and metadata.

flomonster commented on editoast/src/views/rolling_stocks.rs:
```diff
@@ -39,12 +39,15 @@ pub enum RollingStockError {
     NameAlreadyUsed { name: String },
+    #[error("RollingStock '{name}' is locked")]
+    #[editoast_error(status = 400)]
+    RollingStockIsLocked { name: String },
```
We should rather return the id of the rolling stock than its name.

```suggestion
    RollingStockIsLocked { rolling_stock_id: i64 },
```

flomonster commented on editoast/src/views/rolling_stocks.rs:
```diff
@@ -85,6 +89,7 @@ impl From<RollingStockForm> for RollingStockModel {
         RollingStockModel {
             name: Some(rolling_stock.name),
             version: Some(rolling_stock.version),
+            locked: Some(rolling_stock.locked.unwrap_or(false)),
```
As we discussed, the locked field should better be a query param than a field of the rolling stock payload.

Erashin replied:
I don't really know what I did, but i'm pretty sure i put the locked as a query params in the infra model in my other pr, instead of here, on the rolling stock. I'll take care of that.

flomonster commented on editoast/src/views/rolling_stocks.rs:
```diff
@@ -170,42 +177,71 @@ async fn update(
+) -> Result<HttpResponse> {
+    let data = data.into_inner();
+    let rolling_stock_id = rolling_stock_id.into_inner();
+    let mut rolling_stock = retrieve_existing_rolling_stock(&db_pool, rolling_stock_id).await?;
```
You don't need to retrieve the rolling stock to update one of its fields. You can build the payload with its id and `locked` value and the model system will take care of the rest.

Erashin replied:
Correct. Since we don't need to check if the rolling_stock was previously locked or not in this case, this is completely irrelevant.

flomonster commented on front/yarn.lock:
I think this change was a mistake.

flomonster commented on python/osrd_schemas/osrd_schemas/rolling_stock.py:
```diff
@@ -119,6 +119,7 @@ class RollingStock(BaseModel, extra=Extra.forbid):
 
     version: Literal[RAILJSON_ROLLING_STOCK_VERSION] = Field(default=RAILJSON_ROLLING_STOCK_VERSION)
     name: constr(max_length=255)
+    locked: bool = Field(default=False, description="Whether the rolling stock can be edited/deleted or not.")
```
The schema doesn't need to be change.

flomonster reviewed (changes requested):
Thanks for your PR. The main suggestion is to not change the rolling stock schema and consider the locked attribute as metadata.

Erashin commented on editoast/src/tests/example_rolling_stock_1.json:
```diff
@@ -1,5 +1,6 @@
 {
     "version": "3.1",
+    "locked": false,
```
As we now have a distinct schema, RollingStockFull, for the front endpoints, and since this is used for endpoint testing, it makes sense to me to add locked to the examples (just like metadata is also in the examples). If you don't agree, i can always post add it in fixtures.rs.

leovalais commented on editoast/src/core/pathfinding.rs:
```diff
@@ -15,7 +15,7 @@ pub struct PathfindingRequest {
     infra: i64,
     expected_version: String,
     waypoints: PathfindingWaypoints,
-    rolling_stocks: Vec<RollingStockForm>,
```
If I understand correctly you changed that because you don't want to send a `locked` key that Core doesn't expect. But by replacing it by `RollingStock` you send the `id`, that Core doesn't expect either.

It works for now because Core simply ignores unknown keys but relying on that is extremely fragile imho.

My suggestion would be to extract the rolling stock fields that editoast and core share (ie: everything in `RollingStockForm` but `locked` and `metadata`) and move them in another struct (let's say `RollingStockCommon` but you can change the name), and include it in both `RollingStock` and `RollingStockForm`:

```rust
struct RollingStockForm {
    #[serde(flatten)]
    pub common: RollingStockCommon,
    pub locked: Option<bool>,
    pub metadata: RollingStockMetadata,
}

struct RollingStock {
    pub id: i64,
    #[serde(flatten)]
    pub common: RollingStockCommon,
}
```

Erashin replied:
Wouldn't it be simpler to just move `id` to `RollingStockFull` ? i.e.:
```
struct RollingStockFull {
    pub id: i64,
    #[serde(flatten)]
    pub rolling_stock: RollingStock,
    pub locked: Option<bool>,
    pub metadata: RollingStockMetadata,
}

struct RollingStock {
    ... (without id)
}
```

leovalais replied:
Indeed, since the views now all work with RollingStockFull. But then I think the naming is not ideal, what dot you think about something like that?

```rust
struct RollingStockCommon { // ex RollingStock
    ... (without id and metadata)
}

struct RollingStock { // ex RollingStockFull
    pub id: i64,
    #[serde(flatten)]
    pub common: RollingStockCommon,
    pub locked: Option<bool>,
    pub metadata: RollingStockMetadata,
}

struct RollingStockForm {
    #[serde(flatten)]
    pub common: RollingStockCommon,
    pub locked: Option<bool>,
    pub metadata: RollingStockMetadata,
}
```

Erashin replied:
Yup that works ;)

leovalais commented on editoast/src/views/rolling_stocks.rs:
```diff
@@ -158,10 +135,32 @@ impl RollingStockForm {
     }
 }
 
+impl RollingStockLockedUpdateForm {
```
Place the impl block near the struct definition when possible.

leovalais commented on editoast/src/views/rolling_stocks.rs:
```diff
@@ -248,6 +275,26 @@ async fn create_livery(
+    db_pool: &Data<DbPool>,
+    rolling_stock_id: i64,
+) -> Result<RollingStockModel> {
+    match RollingStockModel::retrieve(db_pool.clone(), rolling_stock_id).await? {
```
Fyi you can can use `Option::ok_or` here.

leovalais commented on editoast/src/views/rolling_stocks.rs:
```diff
@@ -382,20 +429,69 @@ pub mod tests {
+        let patch_body: JsonValue = read_body_json(patch_response).await;
+        let is_locked_error: InternalError =
+            RollingStockError::RollingStockIsLocked { rolling_stock_id }.into();
+        assert_eq!(to_value(&is_locked_error).unwrap(), patch_body);
```
There is a couple macro defs that might make such assertions more concise and avoid `JsonValue` here:

https://github.com/DGEXSolutions/osrd/blob/c8a1770ca54769c13e51c60931707d364e1ec781/editoast/src/views/mod.rs#L93-L126

and find their usage here:

https://github.com/DGEXSolutions/osrd/blob/c8a1770ca54769c13e51c60931707d364e1ec781/editoast/src/views/pathfinding/mod.rs#L519

Erashin replied:
Thanks for that, makes testing errors a whole lot easier.

leovalais reviewed (changes requested):
Thanks! I suggested a few changes:

leovalais reviewed (approved):
LGTM! Manually tested PATCH /rolling_stock/n/locked

flomonster commented on python/osrd_schemas/osrd_schemas/rolling_stock.py:
```diff
@@ -213,7 +213,6 @@ class RollingStock(BaseModel, extra=Extra.forbid):
     mass: PositiveFloat = Field(description="The mass of the train, in kg")
     rolling_resistance: RollingResistance = Field(description="The formula to use to compute rolling resistance")
     loading_gauge: LoadingGaugeType
-    metadata: Mapping[str, str] = Field(description="Properties used in the frontend to display the rolling stock")
```
This shouldn't be removed. 
We should add the locked field.

flomonster reviewed (approved):
Thanks for your PR. LGTM (you just need to adapt the `osrd_schemas/rolling_stock.py`).

Erashin committed:
Add locked to rollingstock model to prevent its edition/deletion

editoast:
   - RollingStock => RollingStockCommon containing only the necessary fields for Core. RollingStock now contains a RollingStockCommon and the necessary fields for front.
   - add locked to rolling_stock && light_rolling_stock models.
   - add query param to POST to set initial rolling_stock locked value. Add control to PATCH && DELETE rolling_stock endpoints to prevent it when locked is true. Add endpoint to edit the locked value of a rolling_stock.
python: api/osrd_schemas: add locked to rolling_stock model.
